File: setup_structure.js
const fs = require('fs');
const path = require('path');

const structure = {
  'src/auth/providers/google': ['oauth_client.py', 'callback_handler.py', 'token_manager.py', '__init__.py'],
  'src/auth/providers/entra': ['msal_config.py', 'enterprise_sso.py', 'conditional_access.py', '__init__.py'],
  'src/auth/mfa/google-mfa': ['enforcement.py', '__init__.py'],
  'src/auth/mfa/ms-authenticator': ['enforcement.py', '__init__.py'],
  'src/auth/mfa/grace-period': ['manager.py', '__init__.py'],
  'src/auth/session': ['session_manager.py', 'timeout_detector.py', 'reauth_trigger.py', '__init__.py'],
  'src/auth/cha': ['continuous_auth.py', 'integrity_validator.py', '__init__.py'],
  'src/roles/student/onboarding': ['invite_validation.py', 'flow.py', '__init__.py'],
  'src/roles/student/permissions': ['rbac.py', '__init__.py'],
  'src/roles/student/error-handlers': ['recovery.py', '__init__.py'],
  'src/roles/administrator/onboarding': ['enrollment.py', 'flow.py', '__init__.py'],
  'src/roles/administrator/permissions': ['rbac.py', '__init__.py'],
  'src/roles/administrator/error-handlers': ['recovery.py', '__init__.py'],
  'src/roles/administrator/pam': ['jit_elevation.py', 'session_recording.py', 'auto_revocation.py', '__init__.py'],
  'src/roles/donor/onboarding': ['donation_form.py', 'flow.py', '__init__.py'],
  'src/roles/donor/permissions': ['rbac.py', '__init__.py'],
  'src/roles/donor/error-handlers': ['recovery.py', '__init__.py'],
  'src/roles/donor/guest-flow': ['guest_donation.py', '__init__.py'],
  'src/roles/volunteer/onboarding': ['vetting.py', 'flow.py', '__init__.py'],
  'src/roles/volunteer/permissions': ['rbac.py', '__init__.py'],
  'src/roles/volunteer/error-handlers': ['recovery.py', '__init__.py'],
  'src/roles/volunteer/class-scope': ['scope_enforcer.py', '__init__.py'],
  'src/platform/community': ['forum.py', 'learning_resources.py', 'impact_dashboard.py', '__init__.py'],
  'src/platform/admin-portal': ['dashboard.py', 'user_management.py', 'role_assignment.py', 'volunteer_vetting.py', 'invite_codes.py', '__init__.py'],
  'src/platform/class-management': ['student_groups.py', 'course_materials.py', '__init__.py'],
  'src/monitoring/audit-log': ['logger.py', 'schema.py', 'storage.py', 'query.py', '__init__.py'],
  'src/monitoring/compliance': ['popia_report.py', 'gdpr_report.py', '__init__.py'],
  'src/monitoring/deprovisioning': ['offboarding.py', 'role_change.py', 'access_cleanup.py', '__init__.py'],
  'infrastructure/identity': ['idp_config.tf', 'mfa_config.tf', 'README.md'],
  'infrastructure/networking': ['api_gateway.tf', 'security_groups.tf', 'README.md'],
  'infrastructure/environments/staging': ['variables.tf', 'terraform.tfvars.example'],
  'infrastructure/environments/production': ['variables.tf', 'terraform.tfvars.example'],
  'config/rbac-policies': ['student.json', 'administrator.json', 'donor.json', 'volunteer.json', 'README.md'],
  'config/session-policies': ['student.json', 'administrator.json', 'donor.json', 'volunteer.json', 'README.md'],
  'config/mfa-policies': ['student.json', 'administrator.json', 'donor.json', 'volunteer.json', 'README.md'],
  'docs': ['architecture.md', 'discovery.md', 'runbooks.md', 'incident_response.md', 'README.md'],
  'scripts': ['access_review.py', 'bulk_provision.py', 'audit_export.py', 'deprovision.py', 'README.md']
};

Object.entries(structure).forEach(([dir, files]) => {
  const fullPath = path.join(__dirname, dir);
  fs.mkdirSync(fullPath, { recursive: true });
  files.forEach(file => {
    fs.writeFileSync(path.join(fullPath, file), '');
  });
});

fs.writeFileSync(path.join(__dirname, 'src/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/auth/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/auth/providers/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/auth/mfa/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/roles/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/roles/student/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/roles/administrator/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/roles/donor/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/roles/volunteer/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/platform/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'src/monitoring/__init__.py'), '');
fs.writeFileSync(path.join(__dirname, 'infrastructure/README.md'), '');
fs.writeFileSync(path.join(__dirname, 'infrastructure/environments/README.md'), '');
fs.writeFileSync(path.join(__dirname, 'config/README.md'), '');

console.log('Directory structure created successfully!');

File: setup_structure.py
"""
Setup script to create Cloud Heroes Africa directory structure
"""
import os

BASE_DIR = os.path.dirname(os.path.abspath(__file__))

STRUCTURE = {
    'src': {
        'auth': {
            'providers': {
                'google': ['oauth_client.py', 'callback_handler.py', 'token_manager.py', '__init__.py'],
                'entra': ['msal_config.py', 'enterprise_sso.py', 'conditional_access.py', '__init__.py'],
                '__init__.py': None
            },
            'mfa': {
                'google-mfa': ['enforcement.py', '__init__.py'],
                'ms-authenticator': ['enforcement.py', '__init__.py'],
                'grace-period': ['manager.py', '__init__.py'],
                '__init__.py': None
            },
            'session': ['session_manager.py', 'timeout_detector.py', 'reauth_trigger.py', '__init__.py'],
            'cha': ['continuous_auth.py', 'integrity_validator.py', '__init__.py'],
            '__init__.py': None
        },
        'roles': {
            'student': {
                'onboarding': ['invite_validation.py', 'flow.py', '__init__.py'],
                'permissions': ['rbac.py', '__init__.py'],
                'error-handlers': ['recovery.py', '__init__.py'],
                '__init__.py': None
            },
            'administrator': {
                'onboarding': ['enrollment.py', 'flow.py', '__init__.py'],
                'permissions': ['rbac.py', '__init__.py'],
                'error-handlers': ['recovery.py', '__init__.py'],
                'pam': ['jit_elevation.py', 'session_recording.py', 'auto_revocation.py', '__init__.py'],
                '__init__.py': None
            },
            'donor': {
                'onboarding': ['donation_form.py', 'flow.py', '__init__.py'],
                'permissions': ['rbac.py', '__init__.py'],
                'error-handlers': ['recovery.py', '__init__.py'],
                'guest-flow': ['guest_donation.py', '__init__.py'],
                '__init__.py': None
            },
            'volunteer': {
                'onboarding': ['vetting.py', 'flow.py', '__init__.py'],
                'permissions': ['rbac.py', '__init__.py'],
                'error-handlers': ['recovery.py', '__init__.py'],
                'class-scope': ['scope_enforcer.py', '__init__.py'],
                '__init__.py': None
            },
            '__init__.py': None
        },
        'platform': {
            'community': ['forum.py', 'learning_resources.py', 'impact_dashboard.py', '__init__.py'],
            'admin-portal': ['dashboard.py', 'user_management.py', 'role_assignment.py', 'volunteer_vetting.py', 'invite_codes.py', '__init__.py'],
            'class-management': ['student_groups.py', 'course_materials.py', '__init__.py'],
            '__init__.py': None
        },
        'monitoring': {
            'audit-log': ['logger.py', 'schema.py', 'storage.py', 'query.py', '__init__.py'],
            'compliance': ['popia_report.py', 'gdpr_report.py', '__init__.py'],
            'deprovisioning': ['offboarding.py', 'role_change.py', 'access_cleanup.py', '__init__.py'],
            '__init__.py': None
        },
        '__init__.py': None
    },
    'infrastructure': {
        'identity': ['idp_config.tf', 'mfa_config.tf', 'README.md'],
        'networking': ['api_gateway.tf', 'security_groups.tf', 'README.md'],
        'environments': {
            'staging': ['variables.tf', 'terraform.tfvars.example'],
            'production': ['variables.tf', 'terraform.tfvars.example'],
            'README.md': None
        },
        'README.md': None
    },
    'config': {
        'rbac-policies': ['student.json', 'administrator.json', 'donor.json', 'volunteer.json', 'README.md'],
        'session-policies': ['student.json', 'administrator.json', 'donor.json', 'volunteer.json', 'README.md'],
        'mfa-policies': ['student.json', 'administrator.json', 'donor.json', 'volunteer.json', 'README.md'],
        'README.md': None
    },
    'docs': ['architecture.md', 'discovery.md', 'runbooks.md', 'incident_response.md', 'README.md'],
    'scripts': ['access_review.py', 'bulk_provision.py', 'audit_export.py', 'deprovision.py', 'README.md']
}

def create_structure(base_path, structure):
    for name, content in structure.items():
        path = os.path.join(base_path, name)
        
        if isinstance(content, dict):
            os.makedirs(path, exist_ok=True)
            create_structure(path, content)
        elif isinstance(content, list):
            os.makedirs(path, exist_ok=True)
            for file in content:
                file_path = os.path.join(path, file)
                if not os.path.exists(file_path):
                    with open(file_path, 'w') as f:
                        f.write('')
        elif content is None:
            if not os.path.exists(path):
                with open(path, 'w') as f:
                    f.write('')

if __name__ == '__main__':
    create_structure(BASE_DIR, STRUCTURE)
    print('Directory structure created successfully!')
